### PR TITLE
[Autoscaler] simulator placement groups 

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

Depends on #11690

This adds placement group support and some documentation to the autoscaler simulator. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_autoscaling_policy.py [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+        self._setup_simulator()
+
+    def _setup_autoscaler(self):
+        self.runner = MockProcessRunner()
+        self.config = yaml.safe_load(open(self.config_path).read())
+
+        self.provider.create_node(
+            {},
+            {
+                TAG_RAY_NODE_KIND: "head",
```
Comment (AmeerHajAli) [inline python/ray/tests/test_autoscaling_policy.py:169]: Can you use `NODE_KIND_HEAD`?

File: python/ray/tests/test_autoscaling_policy.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+
+    def _get_node_to_run(self, bundle, nodes):
+        for ip, node in nodes.items():
+            if node.bundle_fits(bundle):
+                return ip, node
+        return None, None
+
+    def _schedule_placement_group(self, pg, nodes):
+        # This scheduling algorithm is bad, but it is approximately as bad as
+        # the real placement group scheduler.
```
Comment (AmeerHajAli) [inline python/ray/tests/test_autoscaling_policy.py:241]: lol.

File: python/ray/tests/test_autoscaling_policy.py [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+                new_work_queue.append(work)
+        self.work_queue = new_work_queue
+
+    def _launch_nodes(self):
+        """Launch all queued nodes. Since this will be run serially after
+        `autoscaler.update` there are no race conditions in checking if the
+        queue is empty.
+        """
+        while not self.node_launcher.queue.empty():
+            config, count, node_type = self.node_launcher.queue.get()
```
Comment (AmeerHajAli) [inline python/ray/tests/test_autoscaling_policy.py:317]: Is it relevant at this stage to assert conditions on the number of nodes launched here?

File: python/ray/tests/test_autoscaling_policy.py [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+
+        time = 0
+        while start_count < len(placement_group_requests) and time < 500:
+            time = simulator.step()
+
+        # for ip, node in simulator.ip_to_nodes.items():
+        #     print(f"{ip}: {node.available_resources}, {node.total_resources}")
+        #     pass
+
+        assert time < 200
```
Comment (AmeerHajAli) [inline python/ray/tests/test_autoscaling_policy.py:579]: Why this asserts 200 but in line 572 its 500?

File: python/ray/tests/test_autoscaling_policy.py [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+        task.node = node
+        task.start_time = self.virtual_time
+        end_time = self.virtual_time + task.duration
+        self.event_queue.put(Event(end_time, SIMULATOR_EVENT_TASK_DONE, task))
+        if task.start_callback:
+            task.start_callback()
+        return True
+
+    def schedule(self):
+        # TODO (Alex): Implement a more realistic scheduling algorithm.
```
Comment (AmeerHajAli) [inline python/ray/tests/test_autoscaling_policy.py:295]: By realistic, do you mean basically use the binpacking/resource demand scheduler?

**Review by AmeerHajAli — APPROVED** [review]
LGTM. left a few small comments.
Can you mention what are the follow up PRs?
It will also be nice if you can list what P0s from the design doc this covers.
UPDATE: checked the other PR and essentially this should be all P0s?
Pardon my ignorance, but where do we cover:
- Provider cost (mapping from Node Type -> Instance x Hours)
- Maximum number of machines launched by type

**Comment by ericl** [comment]
Can you rebase this?

File: python/ray/tests/test_autoscaling_policy.py (hunk 5 above)
Comment (wuisawesome) [reply python/ray/tests/test_autoscaling_policy.py:293]: I mostly mean a scheduling policy more similar to raylet

File: python/ray/tests/test_autoscaling_policy.py (hunk 3 above)
Comment (wuisawesome) [reply python/ray/tests/test_autoscaling_policy.py]: i don't think so? if someone really wanted to check on the exact type of node at a given point in time, they should assert it in their test case.

File: python/ray/tests/test_autoscaling_policy.py (hunk 4 above)
Comment (wuisawesome) [reply python/ray/tests/test_autoscaling_policy.py]: i think this has been fixed

**Comment by ericl** [comment]
```
//python/ray/tests:test_autoscaling_policy                              TIMEOUT in 3 out of 3 in 75.0s
  Stats over 3 runs: max = 75.0s, min = 75.0s, avg = 75.0s, dev = 0.0s
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_autoscaling_policy/test.log
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_autoscaling_policy/test_attempts/attempt_1.log
  /home/travis/.cache/bazel/_bazel_travis/b88c129a127452fc94033a29d9f90e20/execroot/com_github_ray_project_ray/bazel-out/k8-opt/testlogs/python/ray/tests/test_autoscaling_policy/test_attempts/attempt_2.log
```

File: python/ray/tests/BUILD [hunk 6]
```diff
@@ -25,7 +26,7 @@ py_test_module_list(
     "test_error_ray_not_initialized.py",
     "test_gcs_fault_tolerance.py",
     "test_iter.py",
-    "test_resource_demand_scheduler.py",
```
Comment (wuisawesome) [inline python/ray/tests/BUILD:28]: Did a tiny bit of reorganizing here since I think "R" comes after "J" in the alphabet :p

**Pull request merged**